From 7794ff4fef286a610c84083a0147414ed43c446c Mon Sep 17 00:00:00 2001
From: Michael W <baddaemon87@gmail.com>
Date: Sat, 13 Jan 2018 12:08:29 +0100
Subject: [PATCH 070/249] wiki: Validate battery

* Add a definition for battery data: removable (True/False),
  capacity [mAh], tech (Li-Ion / Li-Po), fastcharge (True/False)
* Require removable and capacity
* Allow either an array of "model: data" or just "data"
  (valid for all models / just one model in the yml) or "None" (devices
  without battery)
* Make battery field required

Change-Id: Iac4ecd9b6bcb027a631600d3646f9e05de5102a8
---
 _data/devices/A6020.yml         |  2 +-
 _data/devices/FP2.yml           |  2 +-
 _data/devices/P024.yml          |  2 +-
 _data/devices/Z008.yml          |  2 +-
 _data/devices/Z00A.yml          |  2 +-
 _data/devices/Z00D.yml          |  2 +-
 _data/devices/Z00L.yml          |  2 +-
 _data/devices/Z00T.yml          |  2 +-
 _data/devices/addison.yml       |  2 +-
 _data/devices/angler.yml        |  2 +-
 _data/devices/armani.yml        |  2 +-
 _data/devices/athene.yml        |  2 +-
 _data/devices/axon7.yml         |  2 +-
 _data/devices/bacon.yml         |  2 +-
 _data/devices/bullhead.yml      |  2 +-
 _data/devices/cancro.yml        |  4 +++-
 _data/devices/capricorn.yml     |  2 +-
 _data/devices/chagallwifi.yml   |  2 +-
 _data/devices/che10.yml         |  4 +++-
 _data/devices/cheeseburger.yml  |  2 +-
 _data/devices/cherry.yml        |  4 +++-
 _data/devices/clark.yml         |  2 +-
 _data/devices/condor.yml        |  2 +-
 _data/devices/crackling.yml     |  2 +-
 _data/devices/d2att.yml         |  2 +-
 _data/devices/d2spr.yml         |  2 +-
 _data/devices/d2tmo.yml         |  2 +-
 _data/devices/d2vzw.yml         |  2 +-
 _data/devices/d800.yml          |  2 +-
 _data/devices/d801.yml          |  2 +-
 _data/devices/d802.yml          |  2 +-
 _data/devices/d803.yml          |  2 +-
 _data/devices/d850.yml          |  2 +-
 _data/devices/d851.yml          |  2 +-
 _data/devices/d852.yml          |  2 +-
 _data/devices/d855.yml          |  2 +-
 _data/devices/deb.yml           |  2 +-
 _data/devices/dogo.yml          |  2 +-
 _data/devices/dragon.yml        |  2 +-
 _data/devices/espresso3g.yml    |  4 +++-
 _data/devices/espressowifi.yml  |  4 +++-
 _data/devices/ether.yml         |  2 +-
 _data/devices/f400.yml          |  2 +-
 _data/devices/falcon.yml        |  2 +-
 _data/devices/find7.yml         |  2 +-
 _data/devices/find7s.yml        |  2 +-
 _data/devices/flo.yml           |  2 +-
 _data/devices/flounder.yml      |  2 +-
 _data/devices/flounder_lte.yml  |  2 +-
 _data/devices/foster.yml        |  2 +-
 _data/devices/fugu.yml          |  2 +-
 _data/devices/g2m.yml           |  2 +-
 _data/devices/gemini.yml        |  2 +-
 _data/devices/ghost.yml         |  2 +-
 _data/devices/gohan.yml         |  2 +-
 _data/devices/griffin.yml       |  2 +-
 _data/devices/gtelwifiue.yml    |  2 +-
 _data/devices/gtesqltespr.yml   |  2 +-
 _data/devices/gts210ltexx.yml   |  2 +-
 _data/devices/gts210vewifi.yml  |  2 +-
 _data/devices/gts210wifi.yml    |  2 +-
 _data/devices/gts28vewifi.yml   |  2 +-
 _data/devices/h811.yml          |  2 +-
 _data/devices/h815.yml          |  2 +-
 _data/devices/h830.yml          |  2 +-
 _data/devices/h850.yml          |  2 +-
 _data/devices/h870.yml          |  2 +-
 _data/devices/h910.yml          |  2 +-
 _data/devices/h918.yml          |  2 +-
 _data/devices/ha3g.yml          |  2 +-
 _data/devices/ham.yml           |  2 +-
 _data/devices/hammerhead.yml    |  2 +-
 _data/devices/harpia.yml        |  2 +-
 _data/devices/hayabusa.yml      |  2 +-
 _data/devices/hero2lte.yml      |  2 +-
 _data/devices/herolte.yml       |  2 +-
 _data/devices/hiaeuhl.yml       |  2 +-
 _data/devices/hiaeul.yml        |  2 +-
 _data/devices/himaul.yml        |  2 +-
 _data/devices/himawl.yml        |  2 +-
 _data/devices/hlte.yml          |  2 +-
 _data/devices/hltetmo.yml       |  2 +-
 _data/devices/huashan.yml       |  2 +-
 _data/devices/hydrogen.yml      |  2 +-
 _data/devices/i9100.yml         |  2 +-
 _data/devices/i9300.yml         |  1 +
 _data/devices/i9305.yml         |  2 +-
 _data/devices/ido.yml           |  2 +-
 _data/devices/ivy.yml           |  2 +-
 _data/devices/jag3gds.yml       |  2 +-
 _data/devices/jagnm.yml         |  2 +-
 _data/devices/jalebi.yml        |  2 +-
 _data/devices/jfltevzw.yml      |  2 +-
 _data/devices/jfltexx.yml       |  2 +-
 _data/devices/k3gxx.yml         |  2 +-
 _data/devices/karin.yml         |  2 +-
 _data/devices/karin_windy.yml   |  2 +-
 _data/devices/kccat6.yml        |  2 +-
 _data/devices/kenzo.yml         |  2 +-
 _data/devices/kingdom.yml       |  2 +-
 _data/devices/kipper.yml        |  2 +-
 _data/devices/kiwi.yml          |  2 +-
 _data/devices/klimtwifi.yml     |  2 +-
 _data/devices/klte.yml          |  2 +-
 _data/devices/kltechn.yml       |  2 +-
 _data/devices/kltechnduo.yml    |  2 +-
 _data/devices/klteduos.yml      |  2 +-
 _data/devices/kltedv.yml        |  2 +-
 _data/devices/kltekdi.yml       |  2 +-
 _data/devices/kltekor.yml       |  2 +-
 _data/devices/kltesprsports.yml |  2 +-
 _data/devices/ks01lte.yml       |  2 +-
 _data/devices/lentislte.yml     |  2 +-
 _data/devices/lettuce.yml       |  2 +-
 _data/devices/libra.yml         |  2 +-
 _data/devices/lithium.yml       |  2 +-
 _data/devices/ls990.yml         |  2 +-
 _data/devices/ls997.yml         |  2 +-
 _data/devices/lt02ltespr.yml    |  2 +-
 _data/devices/lt03lte.yml       |  2 +-
 _data/devices/lux.yml           |  2 +-
 _data/devices/m216.yml          |  2 +-
 _data/devices/m7.yml            |  2 +-
 _data/devices/m7vzw.yml         |  2 +-
 _data/devices/m8.yml            |  2 +-
 _data/devices/m8d.yml           |  2 +-
 _data/devices/maguro.yml        |  2 +-
 _data/devices/mako.yml          |  2 +-
 _data/devices/manta.yml         |  2 +-
 _data/devices/maserati.yml      |  2 +-
 _data/devices/merlin.yml        |  2 +-
 _data/devices/mido.yml          |  2 +-
 _data/devices/mint.yml          |  2 +-
 _data/devices/mondrianwifi.yml  |  2 +-
 _data/devices/mt2.yml           |  2 +-
 _data/devices/n1awifi.yml       |  2 +-
 _data/devices/n2awifi.yml       |  2 +-
 _data/devices/n3.yml            |  2 +-
 _data/devices/n5100.yml         |  2 +-
 _data/devices/n5110.yml         |  2 +-
 _data/devices/n5120.yml         |  2 +-
 _data/devices/natrium.yml       |  2 +-
 _data/devices/nicki.yml         |  2 +-
 _data/devices/nx512j.yml        |  2 +-
 _data/devices/odin.yml          |  2 +-
 _data/devices/oneplus2.yml      |  2 +-
 _data/devices/oneplus3.yml      |  4 +++-
 _data/devices/onyx.yml          |  2 +-
 _data/devices/osprey.yml        |  2 +-
 _data/devices/otus.yml          |  2 +-
 _data/devices/paella.yml        |  2 +-
 _data/devices/peach.yml         |  2 +-
 _data/devices/peregrine.yml     |  2 +-
 _data/devices/piccolo.yml       |  2 +-
 _data/devices/pme.yml           |  2 +-
 _data/devices/pollux.yml        |  2 +-
 _data/devices/pollux_windy.yml  |  2 +-
 _data/devices/r5.yml            |  2 +-
 _data/devices/r7plus.yml        |  2 +-
 _data/devices/r7sf.yml          |  2 +-
 _data/devices/roth.yml          |  2 +-
 _data/devices/s2.yml            |  2 +-
 _data/devices/seed.yml          |  2 +-
 _data/devices/serrano3gxx.yml   |  2 +-
 _data/devices/serranodsdd.yml   |  2 +-
 _data/devices/serranoltexx.yml  |  2 +-
 _data/devices/shamu.yml         |  2 +-
 _data/devices/shieldtablet.yml  |  2 +-
 _data/devices/spyder.yml        |  2 +-
 _data/devices/surnia.yml        |  2 +-
 _data/devices/suzuran.yml       |  2 +-
 _data/devices/t0lte.yml         |  2 +-
 _data/devices/t6.yml            |  2 +-
 _data/devices/t6vzw.yml         |  2 +-
 _data/devices/taoshan.yml       |  2 +-
 _data/devices/targa.yml         |  2 +-
 _data/devices/tenshi.yml        |  2 +-
 _data/devices/thea.yml          |  2 +-
 _data/devices/titan.yml         |  2 +-
 _data/devices/tomato.yml        |  2 +-
 _data/devices/toro.yml          |  2 +-
 _data/devices/toroplus.yml      |  2 +-
 _data/devices/tsubasa.yml       |  2 +-
 _data/devices/umts_spyder.yml   |  4 +++-
 _data/devices/us996.yml         |  2 +-
 _data/devices/us997.yml         |  2 +-
 _data/devices/v1awifi.yml       |  2 +-
 _data/devices/v400.yml          |  2 +-
 _data/devices/v410.yml          |  2 +-
 _data/devices/v480.yml          |  2 +-
 _data/devices/v500.yml          |  2 +-
 _data/devices/v521.yml          |  2 +-
 _data/devices/vegetalte.yml     |  2 +-
 _data/devices/victara.yml       |  2 +-
 _data/devices/vs985.yml         |  2 +-
 _data/devices/vs995.yml         |  2 +-
 _data/devices/w5.yml            |  2 +-
 _data/devices/w7.yml            |  2 +-
 _data/devices/wt88047.yml       |  2 +-
 _data/devices/x2.yml            |  2 +-
 _data/devices/xt897.yml         |  2 +-
 _data/devices/yuga.yml          |  2 +-
 _data/devices/zl1.yml           |  2 +-
 _includes/deviceinfo.html       | 21 +++++++++++++++---
 _includes/snippets/battery.html | 15 +++++++++++++
 device_sample/sample.yml        |  4 +++-
 test/schema-06.yml              | 38 +++++++++++++++++++++++++++++++++
 207 files changed, 291 insertions(+), 206 deletions(-)
 create mode 100644 _includes/snippets/battery.html

diff --git a/_data/devices/A6020.yml b/_data/devices/A6020.yml
index da7d9c6..19a56e4 100644
--- a/_data/devices/A6020.yml
+++ b/_data/devices/A6020.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Removable Li-Ion 2750 mAh
+battery: {removable: True, capacity: 2750, tech: 'Li-Ion'}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '13 MP'}
diff --git a/_data/devices/FP2.yml b/_data/devices/FP2.yml
index b6bebba..ebf64c2 100644
--- a/_data/devices/FP2.yml
+++ b/_data/devices/FP2.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 2420 mAh
+battery: {removable: True, capacity: 2420, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'LED', info: '12 MP'}
diff --git a/_data/devices/P024.yml b/_data/devices/P024.yml
index 0d44ae5..1f1a7e7 100644
--- a/_data/devices/P024.yml
+++ b/_data/devices/P024.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Po 4000 mAh
+battery: {removable: False, capacity: 4000, tech: 'Li-Po'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: '', info: '5 / 8 MP'}
diff --git a/_data/devices/Z008.yml b/_data/devices/Z008.yml
index 6f71505..fd23993 100644
--- a/_data/devices/Z008.yml
+++ b/_data/devices/Z008.yml
@@ -1,5 +1,5 @@
 architecture: {cpu: 'x86_64', userspace: 'x86'}
-battery: 3000mAh
+battery: {removable: False, capacity: 3000}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '13 MP'}
diff --git a/_data/devices/Z00A.yml b/_data/devices/Z00A.yml
index bd06255..e5e6090 100644
--- a/_data/devices/Z00A.yml
+++ b/_data/devices/Z00A.yml
@@ -1,5 +1,5 @@
 architecture: {cpu: 'x86_64', userspace: 'x86'}
-battery: 3000mAh
+battery: {removable: False, capacity: 3000}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '13 MP'}
diff --git a/_data/devices/Z00D.yml b/_data/devices/Z00D.yml
index e3c44dc..0a18442 100644
--- a/_data/devices/Z00D.yml
+++ b/_data/devices/Z00D.yml
@@ -1,5 +1,5 @@
 architecture: x86
-battery: 2500mAh
+battery: {removable: False, capacity: 2500}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/Z00L.yml b/_data/devices/Z00L.yml
index 0e30c05..7fe8820 100644
--- a/_data/devices/Z00L.yml
+++ b/_data/devices/Z00L.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: 3000 mAh
+battery: {removable: False, capacity: 3000}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '13 MP'}
diff --git a/_data/devices/Z00T.yml b/_data/devices/Z00T.yml
index 7f6fdd4..a58ab4c 100644
--- a/_data/devices/Z00T.yml
+++ b/_data/devices/Z00T.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: 3000 mAh
+battery: {removable: False, capacity: 3000}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '13 MP'}
diff --git a/_data/devices/addison.yml b/_data/devices/addison.yml
index 0bdf4e1..f240a17 100644
--- a/_data/devices/addison.yml
+++ b/_data/devices/addison.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable 3510 mAh
+battery: {removable: False, capacity: 3510}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'Dual LED', info: '16 MP'}
diff --git a/_data/devices/angler.yml b/_data/devices/angler.yml
index e2d4d68..d3c2e64 100644
--- a/_data/devices/angler.yml
+++ b/_data/devices/angler.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: 3450 mAh, Li-Po non-removable
+battery: {removable: False, capacity: 3450, tech: 'Li-Po'}
 bluetooth: {spec: '4.2', profiles: [A2DP]}
 cameras:
 - {flash: 'Dual LED (dual tone)', info: '12.3 MP, laser autofocus'}
diff --git a/_data/devices/armani.yml b/_data/devices/armani.yml
index c722b8d..3840535 100644
--- a/_data/devices/armani.yml
+++ b/_data/devices/armani.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 2000 mAh
+battery: {removable: True, capacity: 2000, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/athene.yml b/_data/devices/athene.yml
index 806435f..4aa9bb7 100644
--- a/_data/devices/athene.yml
+++ b/_data/devices/athene.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 3000 mAh
+battery: {removable: False, capacity: 3000}
 bluetooth: {spec: '4.1'}
 cameras:
 - {flash: 'Dual LED', info: '16 MP'}
diff --git a/_data/devices/axon7.yml b/_data/devices/axon7.yml
index be29c8e..80d1b85 100644
--- a/_data/devices/axon7.yml
+++ b/_data/devices/axon7.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Ion 3250 mAh
+battery: {removable: False, capacity: 3250, tech: 'Li-Ion'}
 bluetooth: {spec: '4.2', profiles: [A2DP]}
 cameras:
 - {flash: 'Dual LED (dual tone)', info: '20 MP'}
diff --git a/_data/devices/bacon.yml b/_data/devices/bacon.yml
index 0e43e57..61c60fa 100644
--- a/_data/devices/bacon.yml
+++ b/_data/devices/bacon.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 3100 mAh
+battery: {removable: False, capacity: 3100}
 bluetooth: {spec: '4.1'}
 cameras:
 - {flash: 'Dual LED', info: '13 MP  F/2.0'}
diff --git a/_data/devices/bullhead.yml b/_data/devices/bullhead.yml
index 9c576c3..0f13be1 100644
--- a/_data/devices/bullhead.yml
+++ b/_data/devices/bullhead.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Po 2700 mAh
+battery: {removable: False, capacity: 2700, tech: 'Li-Po'}
 bluetooth: {spec: '4.2', profiles: [A2DP]}
 cameras:
 - {flash: 'Dual LED (dual tone)', info: '12.3 MP, laser autofocus'}
diff --git a/_data/devices/cancro.yml b/_data/devices/cancro.yml
index 6a48f95..b73c326 100644
--- a/_data/devices/cancro.yml
+++ b/_data/devices/cancro.yml
@@ -1,5 +1,7 @@
 architecture: arm
-battery: Non-removable Li-Ion  3050 mAh (Mi 3) / 3080 mAh (Mi 4)
+battery:
+- Mi 3: {removable: False, capacity: 3050, tech: 'Li-Ion'}
+- Mi 4: {removable: False, capacity: 3080, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '13 MP Sony IMX135 Exmor RS (Mi 3) / 13 MP Sony IMX214 Exmor RS (Mi 4)'}
diff --git a/_data/devices/capricorn.yml b/_data/devices/capricorn.yml
index 080744a..d19a0d0 100644
--- a/_data/devices/capricorn.yml
+++ b/_data/devices/capricorn.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Po 3200 mAh
+battery: {removable: False, capacity: 3200, tech: 'Li-Po'}
 bluetooth: {spec: '4.2', profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'Dual LED (dual tone)', info: '12 MP'}
diff --git a/_data/devices/chagallwifi.yml b/_data/devices/chagallwifi.yml
index 67bc56b..96b66fa 100644
--- a/_data/devices/chagallwifi.yml
+++ b/_data/devices/chagallwifi.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Ion 7900 mAh
+battery: {removable: False, capacity: 7900, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/che10.yml b/_data/devices/che10.yml
index 2fea0f4..a876ad7 100644
--- a/_data/devices/che10.yml
+++ b/_data/devices/che10.yml
@@ -1,5 +1,7 @@
 architecture: arm
-battery: Li-Ion 2000 / 3000 mAh
+battery:
+- Honor 4X: {removable: False, capacity: 2000, tech: 'Li-Ion'}
+- Honor 4X: {removable: False, capacity: 3000, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'LED', info: '8 / 13 MP'}
diff --git a/_data/devices/cheeseburger.yml b/_data/devices/cheeseburger.yml
index ff66bad..2b87b75 100644
--- a/_data/devices/cheeseburger.yml
+++ b/_data/devices/cheeseburger.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Ion 3300 mAh
+battery: {removable: False, capacity: 3300, tech: 'Li-Ion'}
 bluetooth: {spec: '4.2', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '20 MP'}
diff --git a/_data/devices/cherry.yml b/_data/devices/cherry.yml
index 6be7fea..01e9014 100644
--- a/_data/devices/cherry.yml
+++ b/_data/devices/cherry.yml
@@ -1,5 +1,7 @@
 architecture: arm
-battery: Li-Ion 2000 / 3000 mAh
+battery:
+- Honor 4: {removable: False, capacity: 2000, tech: 'Li-Ion'}
+- Honor 4X: {removable: False, capacity: 3000, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'LED', info: '8 / 13 MP'}
diff --git a/_data/devices/clark.yml b/_data/devices/clark.yml
index 731bbab..01f06d2 100644
--- a/_data/devices/clark.yml
+++ b/_data/devices/clark.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: 3000 mAh
+battery: {removable: False, capacity: 3000}
 bluetooth: {spec: '4.1'}
 cameras:
 - {flash: 'Dual LED', info: '21 MP'}
diff --git a/_data/devices/condor.yml b/_data/devices/condor.yml
index fab53a8..5558a40 100644
--- a/_data/devices/condor.yml
+++ b/_data/devices/condor.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 1980 mAh
+battery: {removable: False, capacity: 1980}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: '', info: '5 MP'}
diff --git a/_data/devices/crackling.yml b/_data/devices/crackling.yml
index 8eef00c..7b149f4 100644
--- a/_data/devices/crackling.yml
+++ b/_data/devices/crackling.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Removable 2500 mAh
+battery: {removable: True, capacity: 2500}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'Dual LED', info: '13 MP'}
diff --git a/_data/devices/d2att.yml b/_data/devices/d2att.yml
index e42f302..785b10b 100644
--- a/_data/devices/d2att.yml
+++ b/_data/devices/d2att.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2100 mAh
+battery: {removable: False, capacity: 2100}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: '', info: '8 MP w/ LED flash'}
diff --git a/_data/devices/d2spr.yml b/_data/devices/d2spr.yml
index 7760d58..df513f4 100644
--- a/_data/devices/d2spr.yml
+++ b/_data/devices/d2spr.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2100 mAh
+battery: {removable: False, capacity: 2100}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: '', info: '8 MP w/ LED flash'}
diff --git a/_data/devices/d2tmo.yml b/_data/devices/d2tmo.yml
index f15085d..77f4d62 100644
--- a/_data/devices/d2tmo.yml
+++ b/_data/devices/d2tmo.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2100 mAh
+battery: {removable: False, capacity: 2100}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: '', info: '8 MP w/ LED flash'}
diff --git a/_data/devices/d2vzw.yml b/_data/devices/d2vzw.yml
index 7b84a60..588ac0e 100644
--- a/_data/devices/d2vzw.yml
+++ b/_data/devices/d2vzw.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2100 mAh
+battery: {removable: False, capacity: 2100}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: '', info: '8 MP w/ LED flash'}
diff --git a/_data/devices/d800.yml b/_data/devices/d800.yml
index 3b7df81..0e87383 100644
--- a/_data/devices/d800.yml
+++ b/_data/devices/d800.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Po 3000 mAh
+battery: {removable: False, capacity: 3000, tech: 'Li-Po'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'LED', info: '13 MP'}
diff --git a/_data/devices/d801.yml b/_data/devices/d801.yml
index 46ae9de..acf038f 100644
--- a/_data/devices/d801.yml
+++ b/_data/devices/d801.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Po 3000 mAh
+battery: {removable: False, capacity: 3000, tech: 'Li-Po'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'LED', info: '13 MP'}
diff --git a/_data/devices/d802.yml b/_data/devices/d802.yml
index f4115df..d5b1f21 100644
--- a/_data/devices/d802.yml
+++ b/_data/devices/d802.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Po 3000 mAh
+battery: {removable: False, capacity: 3000, tech: 'Li-Po'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'LED', info: '13 MP'}
diff --git a/_data/devices/d803.yml b/_data/devices/d803.yml
index d5fa7ac..bf0d639 100644
--- a/_data/devices/d803.yml
+++ b/_data/devices/d803.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Po 3000 mAh
+battery: {removable: False, capacity: 3000, tech: 'Li-Po'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'LED', info: '13 MP'}
diff --git a/_data/devices/d850.yml b/_data/devices/d850.yml
index 2cd1326..b366623 100644
--- a/_data/devices/d850.yml
+++ b/_data/devices/d850.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 3000 mAh
+battery: {removable: True, capacity: 3000, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'Dual LED', info: '13 MP'}
diff --git a/_data/devices/d851.yml b/_data/devices/d851.yml
index 1485462..e16c5b7 100644
--- a/_data/devices/d851.yml
+++ b/_data/devices/d851.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 3000 mAh
+battery: {removable: True, capacity: 3000, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'Dual LED', info: '13 MP'}
diff --git a/_data/devices/d852.yml b/_data/devices/d852.yml
index 26ceb14..de716ce 100644
--- a/_data/devices/d852.yml
+++ b/_data/devices/d852.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 3000 mAh
+battery: {removable: True, capacity: 3000, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'Dual LED', info: '13 MP'}
diff --git a/_data/devices/d855.yml b/_data/devices/d855.yml
index 6a09c3d..f2eb7dc 100644
--- a/_data/devices/d855.yml
+++ b/_data/devices/d855.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 3000 mAh
+battery: {removable: True, capacity: 3000, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'Dual LED', info: '13 MP'}
diff --git a/_data/devices/deb.yml b/_data/devices/deb.yml
index 366cfda..857360a 100644
--- a/_data/devices/deb.yml
+++ b/_data/devices/deb.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 3950 mAh
+battery: {removable: False, capacity: 3950}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: '', info: '5 MP'}
diff --git a/_data/devices/dogo.yml b/_data/devices/dogo.yml
index 9b6a034..f745423 100644
--- a/_data/devices/dogo.yml
+++ b/_data/devices/dogo.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable 2300 mAh
+battery: {removable: True, capacity: 2300}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '13.1MP f/2.4'}
diff --git a/_data/devices/dragon.yml b/_data/devices/dragon.yml
index fb0e064..32947e9 100644
--- a/_data/devices/dragon.yml
+++ b/_data/devices/dragon.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Po 9240 mAh
+battery: {removable: False, capacity: 9240, tech: 'Li-Po'}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: '', info: '8 MP, f/2.4'}
diff --git a/_data/devices/espresso3g.yml b/_data/devices/espresso3g.yml
index 9774ea8..9324135 100644
--- a/_data/devices/espresso3g.yml
+++ b/_data/devices/espresso3g.yml
@@ -1,5 +1,7 @@
 architecture: arm
-battery: Non-removable Li-Ion 4000 mAh / Li-Po 7000 mAh
+battery:
+- Galaxy Tab 2 7.0: {removable: False, capacity: 4000, tech: 'Li-Ion'}
+- Galaxy Tab 2 10.1: {removable: False, capacity: 7000, tech: 'Li-Po'}
 bluetooth: {spec: '3.0', profiles: [A2DP]}
 cameras:
 - {flash: '', info: '15 MP '}
diff --git a/_data/devices/espressowifi.yml b/_data/devices/espressowifi.yml
index 7da3351..c7fc985 100644
--- a/_data/devices/espressowifi.yml
+++ b/_data/devices/espressowifi.yml
@@ -1,5 +1,7 @@
 architecture: arm
-battery: Non-removable Li-Ion 4000 mAh / Li-Po 7000 mAh
+battery:
+- Galaxy Tab 2 7.0: {removable: False, capacity: 4000, tech: 'Li-Ion'}
+- Galaxy Tab 2 10.1: {removable: False, capacity: 7000, tech: 'Li-Po'}
 bluetooth: {spec: '3.0', profiles: [A2DP]}
 cameras:
 - {flash: '', info: '15 MP '}
diff --git a/_data/devices/ether.yml b/_data/devices/ether.yml
index 22dcf2a..dcc70d4 100644
--- a/_data/devices/ether.yml
+++ b/_data/devices/ether.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Ion 2680 mAh
+battery: {removable: False, capacity: 2680, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'LED (dual-tone)', info: '13 MP'}
diff --git a/_data/devices/f400.yml b/_data/devices/f400.yml
index 22727dc..9d39366 100644
--- a/_data/devices/f400.yml
+++ b/_data/devices/f400.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 3000 mAh
+battery: {removable: True, capacity: 3000, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'Dual LED', info: '13 MP'}
diff --git a/_data/devices/falcon.yml b/_data/devices/falcon.yml
index 5df99d5..1881c79 100644
--- a/_data/devices/falcon.yml
+++ b/_data/devices/falcon.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2070 mAh
+battery: {removable: False, capacity: 2070}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '5 MP'}
diff --git a/_data/devices/find7.yml b/_data/devices/find7.yml
index 2910704..0310f49 100644
--- a/_data/devices/find7.yml
+++ b/_data/devices/find7.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable 2800 mAh Li-Po with Rapid Charge
+battery: {removable: True, capacity: 2800, tech: 'Li-Po', fastcharge: True}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'Dual LED', info: '13 MP'}
diff --git a/_data/devices/find7s.yml b/_data/devices/find7s.yml
index d29a539..3d021fe 100644
--- a/_data/devices/find7s.yml
+++ b/_data/devices/find7s.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable 3000 mAh Li-Po with Rapid Charge
+battery: {removable: True, capacity: 3000, tech: 'Li-Po', fastcharge: True}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'Dual LED', info: '13 MP'}
diff --git a/_data/devices/flo.yml b/_data/devices/flo.yml
index 2ccf14a..5765953 100644
--- a/_data/devices/flo.yml
+++ b/_data/devices/flo.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 3950 mAh
+battery: {removable: False, capacity: 3950}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: '', info: '5 MP'}
diff --git a/_data/devices/flounder.yml b/_data/devices/flounder.yml
index abecc13..b64a1a8 100644
--- a/_data/devices/flounder.yml
+++ b/_data/devices/flounder.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Po 6700 mAh
+battery: {removable: False, capacity: 6700, tech: 'Li-Po'}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/flounder_lte.yml b/_data/devices/flounder_lte.yml
index d440d01..48a27d1 100644
--- a/_data/devices/flounder_lte.yml
+++ b/_data/devices/flounder_lte.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Po 6700 mAh
+battery: {removable: False, capacity: 6700, tech: 'Li-Po'}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/foster.yml b/_data/devices/foster.yml
index c849bf6..3727bbd 100644
--- a/_data/devices/foster.yml
+++ b/_data/devices/foster.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery:
+battery: None
 bluetooth: {spec: '4.1'}
 cameras: []
 carrier:
diff --git a/_data/devices/fugu.yml b/_data/devices/fugu.yml
index 8ce1531..5c5a245 100644
--- a/_data/devices/fugu.yml
+++ b/_data/devices/fugu.yml
@@ -1,5 +1,5 @@
 architecture: {cpu: 'x86_64', userspace: 'x86'}
-battery:
+battery: None
 bluetooth: {spec: '4.1'}
 cameras: []
 carrier:
diff --git a/_data/devices/g2m.yml b/_data/devices/g2m.yml
index 921f9c1..31687b6 100644
--- a/_data/devices/g2m.yml
+++ b/_data/devices/g2m.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 2440 mAh
+battery: {removable: True, capacity: 2440, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/gemini.yml b/_data/devices/gemini.yml
index 14153e9..59087ab 100644
--- a/_data/devices/gemini.yml
+++ b/_data/devices/gemini.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Po 3000 mAh
+battery: {removable: False, capacity: 3000, tech: 'Li-Po'}
 bluetooth: {spec: '4.2', profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'Dual LED (dual tone)', info: '16 MP'}
diff --git a/_data/devices/ghost.yml b/_data/devices/ghost.yml
index 9e95230..1c8fc4d 100644
--- a/_data/devices/ghost.yml
+++ b/_data/devices/ghost.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Ion 2200 mAh
+battery: {removable: False, capacity: 2200, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '10 MP'}
diff --git a/_data/devices/gohan.yml b/_data/devices/gohan.yml
index be3854a..6109a20 100644
--- a/_data/devices/gohan.yml
+++ b/_data/devices/gohan.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Po 3100 mAh
+battery: {removable: False, capacity: 3100, tech: 'Li-Po'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'Dual LED', info: '16 MP'}
diff --git a/_data/devices/griffin.yml b/_data/devices/griffin.yml
index 976b73d..ac1222f 100644
--- a/_data/devices/griffin.yml
+++ b/_data/devices/griffin.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Ion 2600 mAh
+battery: {removable: False, capacity: 2600, tech: 'Li-Ion'}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: 'Dual LED (dual tone)', info: '13 MP, laser autofocus, OIS'}
diff --git a/_data/devices/gtelwifiue.yml b/_data/devices/gtelwifiue.yml
index fc4fb25..fab6dae 100644
--- a/_data/devices/gtelwifiue.yml
+++ b/_data/devices/gtelwifiue.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 7300 mAh
+battery: {removable: True, capacity: 7300, tech: 'Li-Ion'}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: '', info: '5 MP'}
diff --git a/_data/devices/gtesqltespr.yml b/_data/devices/gtesqltespr.yml
index 30929dd..cd57d21 100644
--- a/_data/devices/gtesqltespr.yml
+++ b/_data/devices/gtesqltespr.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 5000 mAh
+battery: {removable: True, capacity: 5000, tech: 'Li-Ion'}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: '', info: '5 MP'}
diff --git a/_data/devices/gts210ltexx.yml b/_data/devices/gts210ltexx.yml
index a113e78..e1b1e23 100644
--- a/_data/devices/gts210ltexx.yml
+++ b/_data/devices/gts210ltexx.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Ion 5870 mAh
+battery: {removable: False, capacity: 5870, tech: 'Li-Ion'}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: '', info: '8 MP'}
diff --git a/_data/devices/gts210vewifi.yml b/_data/devices/gts210vewifi.yml
index 26546e1..fd9032a 100644
--- a/_data/devices/gts210vewifi.yml
+++ b/_data/devices/gts210vewifi.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Ion 5870 mAh
+battery: {removable: False, capacity: 5870, tech: 'Li-Ion'}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: '', info: '8 MP'}
diff --git a/_data/devices/gts210wifi.yml b/_data/devices/gts210wifi.yml
index fae9091..d1c417a 100644
--- a/_data/devices/gts210wifi.yml
+++ b/_data/devices/gts210wifi.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Ion 5870 mAh
+battery: {removable: False, capacity: 5870, tech: 'Li-Ion'}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: '', info: '8 MP'}
diff --git a/_data/devices/gts28vewifi.yml b/_data/devices/gts28vewifi.yml
index 9bcf285..ecaa721 100644
--- a/_data/devices/gts28vewifi.yml
+++ b/_data/devices/gts28vewifi.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Ion 4000 mAh
+battery: {removable: False, capacity: 4000, tech: 'Li-Ion'}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: '', info: '8 MP'}
diff --git a/_data/devices/h811.yml b/_data/devices/h811.yml
index 8010ca1..4c3c2cc 100644
--- a/_data/devices/h811.yml
+++ b/_data/devices/h811.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Removable Li-Ion 3000 mAh
+battery: {removable: True, capacity: 3000, tech: 'Li-Ion'}
 bluetooth: {spec: '4.1', profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/_data/devices/h815.yml b/_data/devices/h815.yml
index aac952c..6fbf8b7 100644
--- a/_data/devices/h815.yml
+++ b/_data/devices/h815.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Removable Li-Ion 3000 mAh
+battery: {removable: True, capacity: 3000, tech: 'Li-Ion'}
 bluetooth: {spec: '4.1', profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/_data/devices/h830.yml b/_data/devices/h830.yml
index 30f5d38..33c96cf 100644
--- a/_data/devices/h830.yml
+++ b/_data/devices/h830.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Removable Li-Ion 2800 mAh
+battery: {removable: True, capacity: 2800, tech: 'Li-Ion'}
 bluetooth: {spec: '4.2', profiles: [A2DP + aptX HD]}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/_data/devices/h850.yml b/_data/devices/h850.yml
index a00c689..30f0b93 100644
--- a/_data/devices/h850.yml
+++ b/_data/devices/h850.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Removable Li-Ion 2800 mAh
+battery: {removable: True, capacity: 2800, tech: 'Li-Ion'}
 bluetooth: {spec: '4.2', profiles: [A2DP + aptX HD]}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/_data/devices/h870.yml b/_data/devices/h870.yml
index 57c121f..ea567b1 100644
--- a/_data/devices/h870.yml
+++ b/_data/devices/h870.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Removable Li-Ion 3300 mAh
+battery: {removable: True, capacity: 3300, tech: 'Li-Ion'}
 bluetooth: {spec: '4.2', profiles: [A2DP + aptX HD]}
 cameras:
 - {flash: 'LED', info: '13 MP'}
diff --git a/_data/devices/h910.yml b/_data/devices/h910.yml
index 437f71a..bdb5c27 100644
--- a/_data/devices/h910.yml
+++ b/_data/devices/h910.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Removable Li-Ion 3200 mAh
+battery: {removable: True, capacity: 3200, tech: 'Li-Ion'}
 bluetooth: {spec: '4.2', profiles: [A2DP + aptX HD]}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/_data/devices/h918.yml b/_data/devices/h918.yml
index 8137f6f..d92d06d 100644
--- a/_data/devices/h918.yml
+++ b/_data/devices/h918.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Removable Li-Ion 3200 mAh
+battery: {removable: True, capacity: 3200, tech: 'Li-Ion'}
 bluetooth: {spec: '4.2', profiles: [A2DP + aptX HD]}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/_data/devices/ha3g.yml b/_data/devices/ha3g.yml
index ae3cb94..18df974 100644
--- a/_data/devices/ha3g.yml
+++ b/_data/devices/ha3g.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 3200 mAh
+battery: {removable: True, capacity: 3200, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '13 MP, 1080p Full HD Video Recording @ 60fps'}
diff --git a/_data/devices/ham.yml b/_data/devices/ham.yml
index 7c5705c..9063e18 100644
--- a/_data/devices/ham.yml
+++ b/_data/devices/ham.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Po 4100 mAh
+battery: {removable: False, capacity: 4100, tech: 'Li-Po'}
 bluetooth: {spec: '4.1'}
 cameras:
 - {flash: 'Dual LED', info: '13 MP F/2.2'}
diff --git a/_data/devices/hammerhead.yml b/_data/devices/hammerhead.yml
index 56c6c1f..f84411e 100644
--- a/_data/devices/hammerhead.yml
+++ b/_data/devices/hammerhead.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2300 mAh
+battery: {removable: False, capacity: 2300}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/harpia.yml b/_data/devices/harpia.yml
index dc32d9a..eaafb2e 100644
--- a/_data/devices/harpia.yml
+++ b/_data/devices/harpia.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2800 mAh
+battery: {removable: False, capacity: 2800}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/hayabusa.yml b/_data/devices/hayabusa.yml
index ace5e4d..b86b8f1 100644
--- a/_data/devices/hayabusa.yml
+++ b/_data/devices/hayabusa.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable 1750 mAh
+battery: {removable: True, capacity: 1750}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '13 MP f/2.4'}
diff --git a/_data/devices/hero2lte.yml b/_data/devices/hero2lte.yml
index 6646617..4be61b4 100644
--- a/_data/devices/hero2lte.yml
+++ b/_data/devices/hero2lte.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Ion 3600 mAh
+battery: {removable: False, capacity: 3600, tech: 'Li-Ion'}
 bluetooth: {spec: '4.2', profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'LED', info: '12 MP , f/1.7, 26mm'}
diff --git a/_data/devices/herolte.yml b/_data/devices/herolte.yml
index bc33f1b..d466eb4 100644
--- a/_data/devices/herolte.yml
+++ b/_data/devices/herolte.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Ion 3000 mAh
+battery: {removable: False, capacity: 3000, tech: 'Li-Ion'}
 bluetooth: {spec: '4.2', profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'LED', info: '12 MP , f/1.7, 26mm'}
diff --git a/_data/devices/hiaeuhl.yml b/_data/devices/hiaeuhl.yml
index c4adfe5..f51064d 100644
--- a/_data/devices/hiaeuhl.yml
+++ b/_data/devices/hiaeuhl.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Ion 2150 mAh
+battery: {removable: False, capacity: 2150, tech: 'Li-Ion'}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: 'Dual-LED', info: '13 MP'}
diff --git a/_data/devices/hiaeul.yml b/_data/devices/hiaeul.yml
index b537bf0..2ade418 100644
--- a/_data/devices/hiaeul.yml
+++ b/_data/devices/hiaeul.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Ion 2150 mAh
+battery: {removable: False, capacity: 2150, tech: 'Li-Ion'}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: 'Dual-LED', info: '13 MP'}
diff --git a/_data/devices/himaul.yml b/_data/devices/himaul.yml
index f6748a7..31fea80 100644
--- a/_data/devices/himaul.yml
+++ b/_data/devices/himaul.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: 2840 mAh (non-removable)
+battery: {removable: False, capacity: 2840}
 bluetooth: {spec: '4.1', profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'Dual LED', info: '20 MP'}
diff --git a/_data/devices/himawl.yml b/_data/devices/himawl.yml
index a61231d..6774199 100644
--- a/_data/devices/himawl.yml
+++ b/_data/devices/himawl.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: 2840 mAh (non-removable)
+battery: {removable: False, capacity: 2840}
 bluetooth: {spec: '4.1', profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'Dual LED', info: '20 MP'}
diff --git a/_data/devices/hlte.yml b/_data/devices/hlte.yml
index a141a04..2fc247c 100644
--- a/_data/devices/hlte.yml
+++ b/_data/devices/hlte.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 3200 mAh
+battery: {removable: True, capacity: 3200, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '13 MP, 2160p@30fps, 1080p@60fps'}
diff --git a/_data/devices/hltetmo.yml b/_data/devices/hltetmo.yml
index 1e21365..094826e 100644
--- a/_data/devices/hltetmo.yml
+++ b/_data/devices/hltetmo.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 3200 mAh
+battery: {removable: True, capacity: 3200, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '13 MP, 2160p@30fps, 1080p@60fps'}
diff --git a/_data/devices/huashan.yml b/_data/devices/huashan.yml
index 84449d1..7125059 100644
--- a/_data/devices/huashan.yml
+++ b/_data/devices/huashan.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable 2370 mAh
+battery: {removable: False, capacity: 2370}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8MP f/2.4'}
diff --git a/_data/devices/hydrogen.yml b/_data/devices/hydrogen.yml
index f4306ae..6ae414a 100644
--- a/_data/devices/hydrogen.yml
+++ b/_data/devices/hydrogen.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Ion 4850 mAh
+battery: {removable: False, capacity: 4850, tech: 'Li-Ion'}
 bluetooth: {spec: '4.2', profiles: [A2DP]}
 cameras:
 - {flash: 'LED (dual-tone)', info: '16 MP'}
diff --git a/_data/devices/i9100.yml b/_data/devices/i9100.yml
index db028ef..59aaff6 100644
--- a/_data/devices/i9100.yml
+++ b/_data/devices/i9100.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 1650 mAh
+battery: {removable: False, capacity: 1650}
 bluetooth: {spec: 3.0 + HS}
 cameras:
 - {flash: '', info: '8 MP w/ LED-flash'}
diff --git a/_data/devices/i9300.yml b/_data/devices/i9300.yml
index 55c65d8..f098a4a 100644
--- a/_data/devices/i9300.yml
+++ b/_data/devices/i9300.yml
@@ -1,4 +1,5 @@
 architecture: arm
+battery: {removable: False, capacity: 2100, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/i9305.yml b/_data/devices/i9305.yml
index 8bdea02..302c486 100644
--- a/_data/devices/i9305.yml
+++ b/_data/devices/i9305.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 2100 mAh
+battery: {removable: True, capacity: 2100, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/ido.yml b/_data/devices/ido.yml
index 20e41a4..a887eca 100644
--- a/_data/devices/ido.yml
+++ b/_data/devices/ido.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Ion 4100 mAh
+battery: {removable: False, capacity: 4100, tech: 'Li-Ion'}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '13 MP'}
diff --git a/_data/devices/ivy.yml b/_data/devices/ivy.yml
index 14738dc..374a317 100644
--- a/_data/devices/ivy.yml
+++ b/_data/devices/ivy.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Ion 2930 mAh
+battery: {removable: False, capacity: 2930, tech: 'Li-Ion'}
 bluetooth: {spec: '4.1', profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'LED', info: '20.7 MP F/2.0'}
diff --git a/_data/devices/jag3gds.yml b/_data/devices/jag3gds.yml
index 0177c50..055c179 100644
--- a/_data/devices/jag3gds.yml
+++ b/_data/devices/jag3gds.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 2540 mAh
+battery: {removable: True, capacity: 2540, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/jagnm.yml b/_data/devices/jagnm.yml
index ce20376..1eb693c 100644
--- a/_data/devices/jagnm.yml
+++ b/_data/devices/jagnm.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 2540 mAh
+battery: {removable: True, capacity: 2540, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/jalebi.yml b/_data/devices/jalebi.yml
index e977228..a380d02 100644
--- a/_data/devices/jalebi.yml
+++ b/_data/devices/jalebi.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2000 mAh
+battery: {removable: False, capacity: 2000}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: '', info: '8 MP'}
diff --git a/_data/devices/jfltevzw.yml b/_data/devices/jfltevzw.yml
index b5c39af..d4e6244 100644
--- a/_data/devices/jfltevzw.yml
+++ b/_data/devices/jfltevzw.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 2600 mAh
+battery: {removable: True, capacity: 2600, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'LED', info: '13 MP'}
diff --git a/_data/devices/jfltexx.yml b/_data/devices/jfltexx.yml
index ec71c84..2324521 100644
--- a/_data/devices/jfltexx.yml
+++ b/_data/devices/jfltexx.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 2600 mAh
+battery: {removable: True, capacity: 2600, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'LED', info: '13 MP'}
diff --git a/_data/devices/k3gxx.yml b/_data/devices/k3gxx.yml
index 629cf80..b8c430f 100644
--- a/_data/devices/k3gxx.yml
+++ b/_data/devices/k3gxx.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 2800 mAh
+battery: {removable: True, capacity: 2800, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '16 MP, 1080p Full HD Video Recording @ 60fps'}
diff --git a/_data/devices/karin.yml b/_data/devices/karin.yml
index 636d810..39f430a 100644
--- a/_data/devices/karin.yml
+++ b/_data/devices/karin.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Ion 6000 mAh
+battery: {removable: False, capacity: 6000, tech: 'Li-Ion'}
 bluetooth: {spec: '4.1', profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'LED', info: '8.1 MP F/2.0'}
diff --git a/_data/devices/karin_windy.yml b/_data/devices/karin_windy.yml
index 63821bc..73c270a 100644
--- a/_data/devices/karin_windy.yml
+++ b/_data/devices/karin_windy.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Ion 6000 mAh
+battery: {removable: False, capacity: 6000, tech: 'Li-Ion'}
 bluetooth: {spec: '4.1', profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'LED', info: '8.1 MP F/2.0'}
diff --git a/_data/devices/kccat6.yml b/_data/devices/kccat6.yml
index ee7f9a3..f8cb678 100644
--- a/_data/devices/kccat6.yml
+++ b/_data/devices/kccat6.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2800 mAh Li-ion battery, removable
+battery: {removable: True, capacity: 2800, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/_data/devices/kenzo.yml b/_data/devices/kenzo.yml
index 0e3c1c4..02658ae 100644
--- a/_data/devices/kenzo.yml
+++ b/_data/devices/kenzo.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Po 4050 mAh
+battery: {removable: False, capacity: 4050, tech: 'Li-Po'}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: 'LED (dual-tone)', info: '16 MP'}
diff --git a/_data/devices/kingdom.yml b/_data/devices/kingdom.yml
index 8fe6ee2..c246b50 100644
--- a/_data/devices/kingdom.yml
+++ b/_data/devices/kingdom.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 4000 mAh Li-Po, non-removable
+battery: {removable: False, capacity: 4000, tech: 'Li-Po'}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'Dual LED', info: '16 MP F/2.2, OIS'}
diff --git a/_data/devices/kipper.yml b/_data/devices/kipper.yml
index df7f21c..ab601ba 100644
--- a/_data/devices/kipper.yml
+++ b/_data/devices/kipper.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: 2500 mAh
+battery: {removable: False, capacity: 2500}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '20 MP'}
diff --git a/_data/devices/kiwi.yml b/_data/devices/kiwi.yml
index b7ec00b..f095340 100644
--- a/_data/devices/kiwi.yml
+++ b/_data/devices/kiwi.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: 3000 mAh (non-removable)
+battery: {removable: False, capacity: 3000}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '13 MP, f/2.0, 28mm'}
diff --git a/_data/devices/klimtwifi.yml b/_data/devices/klimtwifi.yml
index 2b286b1..7c66797 100644
--- a/_data/devices/klimtwifi.yml
+++ b/_data/devices/klimtwifi.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Ion 4900 mAh
+battery: {removable: False, capacity: 4900, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/klte.yml b/_data/devices/klte.yml
index 1f90249..8c78965 100644
--- a/_data/devices/klte.yml
+++ b/_data/devices/klte.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2800 mAh Li-ion battery, removable
+battery: {removable: True, capacity: 2800, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/_data/devices/kltechn.yml b/_data/devices/kltechn.yml
index 6af4c65..e521995 100644
--- a/_data/devices/kltechn.yml
+++ b/_data/devices/kltechn.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2800 mAh Li-ion battery, removable
+battery: {removable: True, capacity: 2800, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/_data/devices/kltechnduo.yml b/_data/devices/kltechnduo.yml
index c95665f..9bd22bd 100644
--- a/_data/devices/kltechnduo.yml
+++ b/_data/devices/kltechnduo.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2800 mAh Li-ion battery, removable
+battery: {removable: True, capacity: 2800, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/_data/devices/klteduos.yml b/_data/devices/klteduos.yml
index 679659b..15198ea 100644
--- a/_data/devices/klteduos.yml
+++ b/_data/devices/klteduos.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2800 mAh Li-ion battery, removable
+battery: {removable: True, capacity: 2800, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/_data/devices/kltedv.yml b/_data/devices/kltedv.yml
index c4456d1..2e1382e 100644
--- a/_data/devices/kltedv.yml
+++ b/_data/devices/kltedv.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2800 mAh Li-ion battery, removable
+battery: {removable: True, capacity: 2800, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/_data/devices/kltekdi.yml b/_data/devices/kltekdi.yml
index 6697be0..50cf619 100644
--- a/_data/devices/kltekdi.yml
+++ b/_data/devices/kltekdi.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2800 mAh Li-ion battery, removable
+battery: {removable: True, capacity: 2800, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/_data/devices/kltekor.yml b/_data/devices/kltekor.yml
index d3d9835..12a0ae0 100644
--- a/_data/devices/kltekor.yml
+++ b/_data/devices/kltekor.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2800 mAh Li-ion battery, removable
+battery: {removable: True, capacity: 2800, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/_data/devices/kltesprsports.yml b/_data/devices/kltesprsports.yml
index ec17ecc..b2ea10c 100644
--- a/_data/devices/kltesprsports.yml
+++ b/_data/devices/kltesprsports.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2800 mAh Li-ion battery, removable
+battery: {removable: True, capacity: 2800, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/_data/devices/ks01lte.yml b/_data/devices/ks01lte.yml
index 8b7626f..ceeeefc 100644
--- a/_data/devices/ks01lte.yml
+++ b/_data/devices/ks01lte.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2600 mAh Li-ion, removable
+battery: {removable: True, capacity: 2600, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: '', info: '13 MP, 1080p Full HD Video Recording @ 30fps'}
diff --git a/_data/devices/lentislte.yml b/_data/devices/lentislte.yml
index e12dcad..9f003b7 100644
--- a/_data/devices/lentislte.yml
+++ b/_data/devices/lentislte.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2800 mAh Li-ion battery, removable
+battery: {removable: True, capacity: 2800, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/_data/devices/lettuce.yml b/_data/devices/lettuce.yml
index d6e6bef..210ec39 100644
--- a/_data/devices/lettuce.yml
+++ b/_data/devices/lettuce.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: 2230 mAh
+battery: {removable: False, capacity: 2230}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/libra.yml b/_data/devices/libra.yml
index e6dd695..5a4bf88 100644
--- a/_data/devices/libra.yml
+++ b/_data/devices/libra.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Ion 3080 mAh
+battery: {removable: False, capacity: 3080, tech: 'Li-Ion'}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: 'Dual LED (dual tone)', info: '13 MP, phase detection autofocus'}
diff --git a/_data/devices/lithium.yml b/_data/devices/lithium.yml
index 42b130b..9147a75 100644
--- a/_data/devices/lithium.yml
+++ b/_data/devices/lithium.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Po 4400 mAh
+battery: {removable: False, capacity: 4400, tech: 'Li-Po'}
 bluetooth: {spec: '4.2', profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'Dual LED (dual tone)', info: '16 MP'}
diff --git a/_data/devices/ls990.yml b/_data/devices/ls990.yml
index 6c3048a..d19100c 100644
--- a/_data/devices/ls990.yml
+++ b/_data/devices/ls990.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 3000 mAh
+battery: {removable: True, capacity: 3000, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'Dual LED', info: '13 MP'}
diff --git a/_data/devices/ls997.yml b/_data/devices/ls997.yml
index a5a2936..0024716 100644
--- a/_data/devices/ls997.yml
+++ b/_data/devices/ls997.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Removable Li-Ion 3200 mAh
+battery: {removable: True, capacity: 3200, tech: 'Li-Ion'}
 bluetooth: {spec: '4.2', profiles: [A2DP + aptX HD]}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/_data/devices/lt02ltespr.yml b/_data/devices/lt02ltespr.yml
index 45e65e1..4ad73cf 100644
--- a/_data/devices/lt02ltespr.yml
+++ b/_data/devices/lt02ltespr.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Ion 4000 mAh
+battery: {removable: False, capacity: 4000, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: '', info: '3 MP'}
diff --git a/_data/devices/lt03lte.yml b/_data/devices/lt03lte.yml
index 29bbc8a..3309f27 100644
--- a/_data/devices/lt03lte.yml
+++ b/_data/devices/lt03lte.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 8220 mAh
+battery: {removable: False, capacity: 8220}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/lux.yml b/_data/devices/lux.yml
index b47d43c..7ecc674 100644
--- a/_data/devices/lux.yml
+++ b/_data/devices/lux.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 3630 mAh
+battery: {removable: False, capacity: 3630}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'Dual LED', info: '21 MP'}
diff --git a/_data/devices/m216.yml b/_data/devices/m216.yml
index 511c609..7b3f3c0 100644
--- a/_data/devices/m216.yml
+++ b/_data/devices/m216.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 2300 mAh
+battery: {removable: True, capacity: 2300, tech: 'Li-Ion'}
 bluetooth: {spec: '4.1'}
 cameras:
 - {flash: 'LED', info: '13 MP(LTE) or 8 MP(3G)'}
diff --git a/_data/devices/m7.yml b/_data/devices/m7.yml
index 8667fb8..52262bb 100644
--- a/_data/devices/m7.yml
+++ b/_data/devices/m7.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2300 mAh
+battery: {removable: False, capacity: 2300}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'LED', info: '4 UP'}
diff --git a/_data/devices/m7vzw.yml b/_data/devices/m7vzw.yml
index d95ff3c..f221303 100644
--- a/_data/devices/m7vzw.yml
+++ b/_data/devices/m7vzw.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2300 mAh
+battery: {removable: False, capacity: 2300}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'LED', info: '4 UP'}
diff --git a/_data/devices/m8.yml b/_data/devices/m8.yml
index fb5ab95..658beda 100644
--- a/_data/devices/m8.yml
+++ b/_data/devices/m8.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Po 2600 mAh
+battery: {removable: False, capacity: 2600, tech: 'Li-Po'}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'Dual LED', info: 'Dual 4 MP'}
diff --git a/_data/devices/m8d.yml b/_data/devices/m8d.yml
index f850e9b..e7acd86 100644
--- a/_data/devices/m8d.yml
+++ b/_data/devices/m8d.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Po 2600 mAh
+battery: {removable: False, capacity: 2600, tech: 'Li-Po'}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: 'Dual 4 MP'}
diff --git a/_data/devices/maguro.yml b/_data/devices/maguro.yml
index e8b9290..ff2e827 100644
--- a/_data/devices/maguro.yml
+++ b/_data/devices/maguro.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 1750 mAh
+battery: {removable: True, capacity: 1750, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '5 MP'}
diff --git a/_data/devices/mako.yml b/_data/devices/mako.yml
index 75a7880..33e3319 100644
--- a/_data/devices/mako.yml
+++ b/_data/devices/mako.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 3.8 V 2100 mAh, 8 Wh rechargable Lithium polymer non-removable battery
+battery: {removable: False, capacity: 2100, tech: 'Li-Po'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/manta.yml b/_data/devices/manta.yml
index e3f3f78..e606f1a 100644
--- a/_data/devices/manta.yml
+++ b/_data/devices/manta.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Po 9000 mAh
+battery: {removable: False, capacity: 9000, tech: 'Li-Po'}
 bluetooth: {spec: '3.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '5 MP'}
diff --git a/_data/devices/maserati.yml b/_data/devices/maserati.yml
index e7462df..01898fa 100644
--- a/_data/devices/maserati.yml
+++ b/_data/devices/maserati.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 1785 mAh
+battery: {removable: False, capacity: 1785}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/merlin.yml b/_data/devices/merlin.yml
index ffed778..e43cc4d 100644
--- a/_data/devices/merlin.yml
+++ b/_data/devices/merlin.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2470 mAh
+battery: {removable: False, capacity: 2470}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'Dual LED', info: '13 MP'}
diff --git a/_data/devices/mido.yml b/_data/devices/mido.yml
index 25e78d3..8248594 100644
--- a/_data/devices/mido.yml
+++ b/_data/devices/mido.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Po 4100 mAh
+battery: {removable: False, capacity: 4100, tech: 'Li-Po'}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: 'LED (dual-tone)', info: '13 MP'}
diff --git a/_data/devices/mint.yml b/_data/devices/mint.yml
index 3e0ea45..d17691a 100644
--- a/_data/devices/mint.yml
+++ b/_data/devices/mint.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable 1780 mAh
+battery: {removable: False, capacity: 1780}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '13 MP f/2.4'}
diff --git a/_data/devices/mondrianwifi.yml b/_data/devices/mondrianwifi.yml
index 60e3e9e..d7f0ecd 100644
--- a/_data/devices/mondrianwifi.yml
+++ b/_data/devices/mondrianwifi.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery:
+battery: {removable: False, capacity: 8220, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/mt2.yml b/_data/devices/mt2.yml
index afa011d..ad82bf4 100644
--- a/_data/devices/mt2.yml
+++ b/_data/devices/mt2.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Ion 4050 mAh
+battery: {removable: False, capacity: 4050, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'LED', info: '13 MP'}
diff --git a/_data/devices/n1awifi.yml b/_data/devices/n1awifi.yml
index f08e54a..e598a25 100644
--- a/_data/devices/n1awifi.yml
+++ b/_data/devices/n1awifi.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Po 8220 mAh
+battery: {removable: False, capacity: 8220, tech: 'Li-Po'}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/n2awifi.yml b/_data/devices/n2awifi.yml
index 305b0a7..fcc2dc0 100644
--- a/_data/devices/n2awifi.yml
+++ b/_data/devices/n2awifi.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Ion 8220 mAh
+battery: {removable: False, capacity: 8220, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/n3.yml b/_data/devices/n3.yml
index 594713b..531744e 100644
--- a/_data/devices/n3.yml
+++ b/_data/devices/n3.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable 3000 mAh Li-Po with VOOC Flash Charge
+battery: {removable: False, capacity: 3000, fastcharge: True}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: Dual-mode LED, info: 'Motorized Rotation, 16MP OmniVision OV16825 f/2.2'}
diff --git a/_data/devices/n5100.yml b/_data/devices/n5100.yml
index de626c2..0d927b6 100644
--- a/_data/devices/n5100.yml
+++ b/_data/devices/n5100.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Ion 4600 mAh
+battery: {removable: False, capacity: 4600, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: '', info: '5 MP, autofocus'}
diff --git a/_data/devices/n5110.yml b/_data/devices/n5110.yml
index b331f8b..9c28009 100644
--- a/_data/devices/n5110.yml
+++ b/_data/devices/n5110.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Ion 4600 mAh
+battery: {removable: False, capacity: 4600, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: '', info: '5 MP, autofocus'}
diff --git a/_data/devices/n5120.yml b/_data/devices/n5120.yml
index b9a315e..7ea5c8c 100644
--- a/_data/devices/n5120.yml
+++ b/_data/devices/n5120.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Ion 4600 mAh
+battery: {removable: False, capacity: 4600, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: '', info: '5 MP, autofocus'}
diff --git a/_data/devices/natrium.yml b/_data/devices/natrium.yml
index fa22887..e3249c8 100644
--- a/_data/devices/natrium.yml
+++ b/_data/devices/natrium.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Po 3800 mAh
+battery: {removable: False, capacity: 3800, tech: 'Li-Po'}
 bluetooth: {spec: '4.2', profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'Dual LED (dual tone)', info: '16 MP'}
diff --git a/_data/devices/nicki.yml b/_data/devices/nicki.yml
index 81d3ab0..1062040 100644
--- a/_data/devices/nicki.yml
+++ b/_data/devices/nicki.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Po 1750 mAh
+battery: {removable: True, capacity: 1750, tech: 'Li-Po'}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: '', info: '5 MP'}
diff --git a/_data/devices/nx512j.yml b/_data/devices/nx512j.yml
index c0ab313..90f21e3 100644
--- a/_data/devices/nx512j.yml
+++ b/_data/devices/nx512j.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Ion 2900 mAh
+battery: {removable: False, capacity: 2900, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'LED', info: '15.9 MP'}
diff --git a/_data/devices/odin.yml b/_data/devices/odin.yml
index 8b4cef0..9a1e0e5 100644
--- a/_data/devices/odin.yml
+++ b/_data/devices/odin.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable 2370 mAh
+battery: {removable: False, capacity: 2370}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '13MP f/2.4'}
diff --git a/_data/devices/oneplus2.yml b/_data/devices/oneplus2.yml
index 2b1628e..d093816 100644
--- a/_data/devices/oneplus2.yml
+++ b/_data/devices/oneplus2.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: 3300 mAh
+battery: {removable: False, capacity: 3300}
 bluetooth: {spec: '4.1'}
 cameras:
 - {flash: 'Dual LED', info: '13 MP  F/2.0'}
diff --git a/_data/devices/oneplus3.yml b/_data/devices/oneplus3.yml
index abc038f..a9e5c02 100644
--- a/_data/devices/oneplus3.yml
+++ b/_data/devices/oneplus3.yml
@@ -1,5 +1,7 @@
 architecture: arm64
-battery: Non-removable Li-Ion 3000 mAh (3) / 3400 mA (3T)
+battery:
+- 3: {removable: False, capacity: 3000, tech: 'Li-Ion'}
+- 3T: {removable: False, capacity: 3400, tech: 'Li-Ion'}
 bluetooth: {spec: '4.2', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/_data/devices/onyx.yml b/_data/devices/onyx.yml
index cc24b29..f30296c 100644
--- a/_data/devices/onyx.yml
+++ b/_data/devices/onyx.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Po 2525 mAh
+battery: {removable: False, capacity: 2525, tech: 'Li-Po'}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '13 MP'}
diff --git a/_data/devices/osprey.yml b/_data/devices/osprey.yml
index 72e9d80..8b3b5b3 100644
--- a/_data/devices/osprey.yml
+++ b/_data/devices/osprey.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2470 mAh
+battery: {removable: False, capacity: 2470}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'Dual LED', info: '13 MP'}
diff --git a/_data/devices/otus.yml b/_data/devices/otus.yml
index ebd37f8..cfa5341 100644
--- a/_data/devices/otus.yml
+++ b/_data/devices/otus.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2390 mAh
+battery: {removable: False, capacity: 2390}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: '', info: '5 MP'}
diff --git a/_data/devices/paella.yml b/_data/devices/paella.yml
index c4378b2..c302a54 100644
--- a/_data/devices/paella.yml
+++ b/_data/devices/paella.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Po 2900 mAh
+battery: {removable: False, capacity: 2900, tech: 'Li-Po'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'Dual LED', info: '13 MP'}
diff --git a/_data/devices/peach.yml b/_data/devices/peach.yml
index 0905d69..6484f54 100644
--- a/_data/devices/peach.yml
+++ b/_data/devices/peach.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 2100 mAh
+battery: {removable: True, capacity: 2100, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'Dual LED', info: '8 MP'}
diff --git a/_data/devices/peregrine.yml b/_data/devices/peregrine.yml
index c019e33..3f526bc 100644
--- a/_data/devices/peregrine.yml
+++ b/_data/devices/peregrine.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2070 mAh
+battery: {removable: False, capacity: 2070}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '5 MP'}
diff --git a/_data/devices/piccolo.yml b/_data/devices/piccolo.yml
index 20059ae..b0f5b1a 100644
--- a/_data/devices/piccolo.yml
+++ b/_data/devices/piccolo.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Po 3120 mAh
+battery: {removable: False, capacity: 3120, tech: 'Li-Po'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'Dual LED', info: '13 MP'}
diff --git a/_data/devices/pme.yml b/_data/devices/pme.yml
index b5bfd4f..db3615a 100644
--- a/_data/devices/pme.yml
+++ b/_data/devices/pme.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Ion 3000 mAh
+battery: {removable: False, capacity: 3000, tech: 'Li-Ion'}
 bluetooth: {spec: '4.2', profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'Dual LED', info: '12 MP , f/1.8, 26mm'}
diff --git a/_data/devices/pollux.yml b/_data/devices/pollux.yml
index 7753211..ffce1a5 100644
--- a/_data/devices/pollux.yml
+++ b/_data/devices/pollux.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable 6000 mAh
+battery: {removable: False, capacity: 6000}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8.1MP f/2.4'}
diff --git a/_data/devices/pollux_windy.yml b/_data/devices/pollux_windy.yml
index 9216c8f..6dccab4 100644
--- a/_data/devices/pollux_windy.yml
+++ b/_data/devices/pollux_windy.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable 6000 mAh
+battery: {removable: False, capacity: 6000}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8.1MP f/2.4'}
diff --git a/_data/devices/r5.yml b/_data/devices/r5.yml
index a52bf74..3891805 100644
--- a/_data/devices/r5.yml
+++ b/_data/devices/r5.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Po 2000 mAh
+battery: {removable: False, capacity: 2000, tech: 'Li-Po'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'LED', info: '13 MP'}
diff --git a/_data/devices/r7plus.yml b/_data/devices/r7plus.yml
index 5f44bbb..e131826 100644
--- a/_data/devices/r7plus.yml
+++ b/_data/devices/r7plus.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Po 4100 mAh
+battery: {removable: False, capacity: 4100, tech: 'Li-Po'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'Dual LED', info: '13 MP'}
diff --git a/_data/devices/r7sf.yml b/_data/devices/r7sf.yml
index 5969f39..d8a29c1 100644
--- a/_data/devices/r7sf.yml
+++ b/_data/devices/r7sf.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Po 3070 mAh
+battery: {removable: False, capacity: 3070, tech: 'Li-Po'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'LED', info: '13 MP'}
diff --git a/_data/devices/roth.yml b/_data/devices/roth.yml
index 1a483ab..2d092ce 100644
--- a/_data/devices/roth.yml
+++ b/_data/devices/roth.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 7350 mAh
+battery: {removable: False, capacity: 7350}
 bluetooth: {spec: '3.0'}
 cameras: []
 carrier:
diff --git a/_data/devices/s2.yml b/_data/devices/s2.yml
index 435af61..55492f9 100644
--- a/_data/devices/s2.yml
+++ b/_data/devices/s2.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: 3000 mAh (non-removable)
+battery: {removable: False, capacity: 3000}
 bluetooth: {spec: '4.1', profiles: [A2DP]}
 cameras:
 - {flash: 'Dual LED (dual tone)', info: '16 MP'}
diff --git a/_data/devices/seed.yml b/_data/devices/seed.yml
index 47d6e8d..4fb23fe 100644
--- a/_data/devices/seed.yml
+++ b/_data/devices/seed.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2500 mAh
+battery: {removable: False, capacity: 2500}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'LED', info: '13 MP'}
diff --git a/_data/devices/serrano3gxx.yml b/_data/devices/serrano3gxx.yml
index 3ed8be2..204efe6 100644
--- a/_data/devices/serrano3gxx.yml
+++ b/_data/devices/serrano3gxx.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 1900 mAh
+battery: {removable: True, capacity: 1900, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8 MP, 1080p Full HD Video Recording @ 30fps'}
diff --git a/_data/devices/serranodsdd.yml b/_data/devices/serranodsdd.yml
index 5d90df8..8292a3d 100644
--- a/_data/devices/serranodsdd.yml
+++ b/_data/devices/serranodsdd.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 1900 mAh
+battery: {removable: True, capacity: 1900, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8 MP, 1080p Full HD Video Recording @ 30fps'}
diff --git a/_data/devices/serranoltexx.yml b/_data/devices/serranoltexx.yml
index 6478753..3a2431c 100644
--- a/_data/devices/serranoltexx.yml
+++ b/_data/devices/serranoltexx.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 1900 mAh
+battery: {removable: True, capacity: 1900, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8 MP, 1080p Full HD Video Recording @ 30fps'}
diff --git a/_data/devices/shamu.yml b/_data/devices/shamu.yml
index ea6ae94..f1d2544 100644
--- a/_data/devices/shamu.yml
+++ b/_data/devices/shamu.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Po 3220 mAh
+battery: {removable: False, capacity: 3220, tech: 'Li-Po'}
 bluetooth: {spec: '4.1'}
 cameras:
 - {flash: 'Dual LED', info: '13 MP'}
diff --git a/_data/devices/shieldtablet.yml b/_data/devices/shieldtablet.yml
index c48eb47..10445f8 100644
--- a/_data/devices/shieldtablet.yml
+++ b/_data/devices/shieldtablet.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 5197 mAh
+battery: {removable: False, capacity: 5197}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: '', info: '5 MP'}
diff --git a/_data/devices/spyder.yml b/_data/devices/spyder.yml
index b71fcc5..3d224d8 100644
--- a/_data/devices/spyder.yml
+++ b/_data/devices/spyder.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 1780 mAh
+battery: {removable: False, capacity: 1780}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/surnia.yml b/_data/devices/surnia.yml
index 5347507..72263cb 100644
--- a/_data/devices/surnia.yml
+++ b/_data/devices/surnia.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2390 mAh
+battery: {removable: False, capacity: 2390}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: '', info: '5 MP'}
diff --git a/_data/devices/suzuran.yml b/_data/devices/suzuran.yml
index 1a0774a..3261953 100644
--- a/_data/devices/suzuran.yml
+++ b/_data/devices/suzuran.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Ion 2700 mAh
+battery: {removable: False, capacity: 2700, tech: 'Li-Ion'}
 bluetooth: {spec: '4.1', profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'LED', info: '23 MP f/2.0'}
diff --git a/_data/devices/t0lte.yml b/_data/devices/t0lte.yml
index 206f170..ec0a56e 100644
--- a/_data/devices/t0lte.yml
+++ b/_data/devices/t0lte.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 3100 mAh
+battery: {removable: True, capacity: 3100, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/t6.yml b/_data/devices/t6.yml
index 11ac79b..c687833 100644
--- a/_data/devices/t6.yml
+++ b/_data/devices/t6.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable 3300 mAh
+battery: {removable: False, capacity: 3300}
 bluetooth: {spec: '4.0', profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'LED', info: '4 UP'}
diff --git a/_data/devices/t6vzw.yml b/_data/devices/t6vzw.yml
index 04618e5..dfcf541 100644
--- a/_data/devices/t6vzw.yml
+++ b/_data/devices/t6vzw.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable 3300 mAh
+battery: {removable: False, capacity: 3300}
 bluetooth: {spec: '4.0', profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'LED', info: '4 UP'}
diff --git a/_data/devices/taoshan.yml b/_data/devices/taoshan.yml
index a526304..466e9a5 100644
--- a/_data/devices/taoshan.yml
+++ b/_data/devices/taoshan.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 1750 mAh
+battery: {removable: True, capacity: 1750, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8 MP f/2.4'}
diff --git a/_data/devices/targa.yml b/_data/devices/targa.yml
index f415fd4..1761bd1 100644
--- a/_data/devices/targa.yml
+++ b/_data/devices/targa.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 1735 mAh
+battery: {removable: False, capacity: 1735}
 bluetooth: {spec: 2.1 + EDR, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/tenshi.yml b/_data/devices/tenshi.yml
index dddf302..7285ff9 100644
--- a/_data/devices/tenshi.yml
+++ b/_data/devices/tenshi.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Po 3050 mAh
+battery: {removable: False, capacity: 3050, tech: 'Li-Po'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'Dual LED', info: '16 MP'}
diff --git a/_data/devices/thea.yml b/_data/devices/thea.yml
index 38347aa..6c9886d 100644
--- a/_data/devices/thea.yml
+++ b/_data/devices/thea.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2390 mAh
+battery: {removable: False, capacity: 2390}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/titan.yml b/_data/devices/titan.yml
index 4f2a81b..7257c32 100644
--- a/_data/devices/titan.yml
+++ b/_data/devices/titan.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2070 mAh
+battery: {removable: False, capacity: 2070}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/tomato.yml b/_data/devices/tomato.yml
index fc8e7a8..cb21d42 100644
--- a/_data/devices/tomato.yml
+++ b/_data/devices/tomato.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: 2500 mAh
+battery: {removable: False, capacity: 2500}
 bluetooth: {spec: '4.1'}
 cameras:
 - {flash: 'LED', info: '13 MP F/2.2'}
diff --git a/_data/devices/toro.yml b/_data/devices/toro.yml
index d87bf8d..0d33b9e 100644
--- a/_data/devices/toro.yml
+++ b/_data/devices/toro.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 1850 mAh
+battery: {removable: True, capacity: 1850, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '5 MP'}
diff --git a/_data/devices/toroplus.yml b/_data/devices/toroplus.yml
index 57ca8bd..b653ee9 100644
--- a/_data/devices/toroplus.yml
+++ b/_data/devices/toroplus.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 1850 mAh
+battery: {removable: True, capacity: 1850, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '5 MP'}
diff --git a/_data/devices/tsubasa.yml b/_data/devices/tsubasa.yml
index f476261..914d35c 100644
--- a/_data/devices/tsubasa.yml
+++ b/_data/devices/tsubasa.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable 1700 mAh
+battery: {removable: True, capacity: 1700}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '13 MP f/2.4'}
diff --git a/_data/devices/umts_spyder.yml b/_data/devices/umts_spyder.yml
index c276ee5..8872830 100644
--- a/_data/devices/umts_spyder.yml
+++ b/_data/devices/umts_spyder.yml
@@ -1,5 +1,7 @@
 architecture: arm
-battery: 1780 mAh / 3300 mAh (MAXX)
+battery:
+- RAZR: {removable: False, capacity: 1780}
+- RAZR MAXX: {removable: False, capacity: 3300}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/us996.yml b/_data/devices/us996.yml
index f578396..5b20ace 100644
--- a/_data/devices/us996.yml
+++ b/_data/devices/us996.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Removable Li-Ion 3200 mAh
+battery: {removable: True, capacity: 3200, tech: 'Li-Ion'}
 bluetooth: {spec: '4.2', profiles: [A2DP + aptX HD]}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/_data/devices/us997.yml b/_data/devices/us997.yml
index 1ce2845..dda7568 100644
--- a/_data/devices/us997.yml
+++ b/_data/devices/us997.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Removable Li-Ion 3300 mAh
+battery: {removable: True, capacity: 3300, tech: 'Li-Ion'}
 bluetooth: {spec: '4.2', profiles: [A2DP + aptX HD]}
 cameras:
 - {flash: 'LED', info: '13 MP'}
diff --git a/_data/devices/v1awifi.yml b/_data/devices/v1awifi.yml
index 00c9ef7..44afa68 100644
--- a/_data/devices/v1awifi.yml
+++ b/_data/devices/v1awifi.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Ion 9500 mAh
+battery: {removable: False, capacity: 9500, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/v400.yml b/_data/devices/v400.yml
index 80d96c3..e8a4b92 100644
--- a/_data/devices/v400.yml
+++ b/_data/devices/v400.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Ion 4000 mAh
+battery: {removable: False, capacity: 4000, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: '', info: '3.15 MP'}
diff --git a/_data/devices/v410.yml b/_data/devices/v410.yml
index 95449a6..870b9cb 100644
--- a/_data/devices/v410.yml
+++ b/_data/devices/v410.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Ion 4000 mAh
+battery: {removable: False, capacity: 4000, tech: 'Li-Ion'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: '', info: '5 MP'}
diff --git a/_data/devices/v480.yml b/_data/devices/v480.yml
index 2a20628..be8a148 100644
--- a/_data/devices/v480.yml
+++ b/_data/devices/v480.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Ion 4200 mAh
+battery: {removable: False, capacity: 4200, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: '', info: '5 MP'}
diff --git a/_data/devices/v500.yml b/_data/devices/v500.yml
index 661c166..1bb76ac 100644
--- a/_data/devices/v500.yml
+++ b/_data/devices/v500.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Po 4600 mAh
+battery: {removable: False, capacity: 4600, tech: 'Li-Po'}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: '', info: '5 MP'}
diff --git a/_data/devices/v521.yml b/_data/devices/v521.yml
index e1c819b..d102d5b 100644
--- a/_data/devices/v521.yml
+++ b/_data/devices/v521.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Non-removable Li-Ion 4650 mAh
+battery: {removable: False, capacity: 4650, tech: 'Li-Ion'}
 bluetooth: {spec: '4.2', profiles: [A2DP + aptX HD]}
 cameras:
 - {flash: '', info: '5 MP'}
diff --git a/_data/devices/vegetalte.yml b/_data/devices/vegetalte.yml
index 63ede88..1de0bb9 100644
--- a/_data/devices/vegetalte.yml
+++ b/_data/devices/vegetalte.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable Li-Po 2850 mAh
+battery: {removable: False, capacity: 2850, tech: 'Li-Po'}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'Dual LED', info: '13 MP'}
diff --git a/_data/devices/victara.yml b/_data/devices/victara.yml
index e77c095..121434e 100644
--- a/_data/devices/victara.yml
+++ b/_data/devices/victara.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2300 mAh (non-removable)
+battery: {removable: False, capacity: 2300}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'Dual LED', info: '13 MP, 4128 x 3096 pixels, autofocus'}
diff --git a/_data/devices/vs985.yml b/_data/devices/vs985.yml
index 2c9dd76..04a1e20 100644
--- a/_data/devices/vs985.yml
+++ b/_data/devices/vs985.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 3000 mAh
+battery: {removable: True, capacity: 3000, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'Dual LED', info: '13 MP'}
diff --git a/_data/devices/vs995.yml b/_data/devices/vs995.yml
index d470907..f75a205 100644
--- a/_data/devices/vs995.yml
+++ b/_data/devices/vs995.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: Removable Li-Ion 3200 mAh
+battery: {removable: True, capacity: 3200, tech: 'Li-Ion'}
 bluetooth: {spec: '4.2', profiles: [A2DP + aptX HD]}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/_data/devices/w5.yml b/_data/devices/w5.yml
index 150d3b6..4b409f4 100644
--- a/_data/devices/w5.yml
+++ b/_data/devices/w5.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 2040 mAh
+battery: {removable: False, capacity: 2040}
 bluetooth: {spec: '4.0'}
 cameras:
 - {flash: 'LED', info: '5 MP'}
diff --git a/_data/devices/w7.yml b/_data/devices/w7.yml
index 4606f84..623cb0e 100644
--- a/_data/devices/w7.yml
+++ b/_data/devices/w7.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Ion 2540 mAh
+battery: {removable: True, capacity: 2540, tech: 'Li-Ion'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP + aptX]}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/wt88047.yml b/_data/devices/wt88047.yml
index 4eaa16f..4e7582a 100644
--- a/_data/devices/wt88047.yml
+++ b/_data/devices/wt88047.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Removable Li-Po 2200 mAh
+battery: {removable: True, capacity: 2200, tech: 'Li-Po'}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/x2.yml b/_data/devices/x2.yml
index 096bf42..7d6f3b9 100644
--- a/_data/devices/x2.yml
+++ b/_data/devices/x2.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: 3100 mAh (non-removable)
+battery: {removable: False, capacity: 3100}
 bluetooth: {spec: '4.2', profiles: [A2DP]}
 cameras:
 - {flash: 'Dual LED (dual tone)', info: '21 MP'}
diff --git a/_data/devices/xt897.yml b/_data/devices/xt897.yml
index 9bb14ad..72b481e 100644
--- a/_data/devices/xt897.yml
+++ b/_data/devices/xt897.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: 1785 mAh
+battery: {removable: False, capacity: 1785}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'LED', info: '8 MP'}
diff --git a/_data/devices/yuga.yml b/_data/devices/yuga.yml
index 13cfd77..fba5242 100644
--- a/_data/devices/yuga.yml
+++ b/_data/devices/yuga.yml
@@ -1,5 +1,5 @@
 architecture: arm
-battery: Non-removable 2330 mAh
+battery: {removable: False, capacity: 2330}
 bluetooth: {spec: 4.0 + LE, profiles: [A2DP]}
 cameras:
 - {flash: 'LED', info: '13.1MP f/2.4'}
diff --git a/_data/devices/zl1.yml b/_data/devices/zl1.yml
index 096842c..c8c87cd 100644
--- a/_data/devices/zl1.yml
+++ b/_data/devices/zl1.yml
@@ -1,5 +1,5 @@
 architecture: arm64
-battery: 4070 mAh (non-removable)
+battery: {removable: False, capacity: 4070}
 bluetooth: {spec: '4.2', profiles: [A2DP]}
 cameras:
 - {flash: 'Dual LED (dual tone)', info: '16 MP'}
diff --git a/_includes/deviceinfo.html b/_includes/deviceinfo.html
index 1322a7f..1947f3b 100644
--- a/_includes/deviceinfo.html
+++ b/_includes/deviceinfo.html
@@ -184,12 +184,27 @@ ul, ol {
         <td>{{ device.height }} (h)<br>{{ device.width }} (w)<br>{{ device.depth }} (d)</td>
     </tr>
     {% endif %}
-    {% if device.battery %}
+    {%- if device.battery != 'None' %}
     <tr>
         <td class="header">Battery</td>
-        <td>{{ device.battery }}</td>
+        <td>
+            {% if device.battery.first.size == 1 -%}
+            {%- comment %}The size of "first" will be 1 when there is a list of models{% endcomment -%}
+                {% for model in device.battery -%}
+                    {%- assign modelname = model.first[0] %}
+                    {%- assign battery_data = model.first[1] %}
+                    {% include snippets/battery.html %}
+                    {%- unless forloop.last -%}
+                    <br/>
+                    {%- endunless %}
+                {% endfor -%}
+            {%- else %}
+                {%- assign battery_data = device.battery %}
+                {% include snippets/battery.html %}
+            {%- endif -%}
+        </td>
     </tr>
-    {% endif %}
+    {%- endif %}
     <!-- lineage info -->
     {% if device.maintainers %}
     <tr>
diff --git a/_includes/snippets/battery.html b/_includes/snippets/battery.html
new file mode 100644
index 0000000..e8ebceb
--- /dev/null
+++ b/_includes/snippets/battery.html
@@ -0,0 +1,15 @@
+{%- if modelname %}
+{{ modelname }}:&nbsp;
+{%- endif %}
+{%- if battery_data.removable -%}
+Removable
+{%- else -%}
+Non-removable
+{%- endif -%}&nbsp;
+{%- if battery_data.tech -%}
+    {{ battery_data.tech }}&nbsp;
+{%- endif -%}
+{{ battery_data.capacity }} mAh
+{%- if battery_data.fastcharge %}
+    &nbsp;with fast charging
+{%- endif %}
diff --git a/device_sample/sample.yml b/device_sample/sample.yml
index 4891eaf..a27b557 100644
--- a/device_sample/sample.yml
+++ b/device_sample/sample.yml
@@ -1,5 +1,7 @@
 architecture: {cpu: 'arm64', userspace: 'arm'}
-battery: Removable Li-Ion 2800 mAh
+battery:
+- A0001: {removable: True, capacity: 2800, tech: 'Li-Ion'}
+- SM-6900F: {removable: True, capacity: 3000, tech: 'Li-Po', fastcharge: True}
 bluetooth: {spec: 4.0 + LE}
 cameras:
 - {flash: 'LED', info: '16 MP'}
diff --git a/test/schema-06.yml b/test/schema-06.yml
index ee7fb19..6cc0d64 100644
--- a/test/schema-06.yml
+++ b/test/schema-06.yml
@@ -37,6 +37,27 @@ definitions:
     - cpu
     - userspace
     type: object
+  battery_data:
+    description: The format required for battery
+    properties:
+      capacity:
+        description: The size of the battery in [mAh]
+        type: number
+      removable:
+        description: Defines whether the battery is removable
+        type: boolean
+      tech:
+        description: The battery technology
+        enum:
+        - Li-Ion
+        - Li-Po
+      fastcharge:
+        description: If the device is capable of fast charging
+        type: boolean
+    required:
+    - capacity
+    - removable
+    type: object
 properties:
   architecture:
     "$id": "/properties/architecture"
@@ -46,6 +67,22 @@ properties:
       "$ref": "#/definitions/valid_architectures"
     - description: An object to provide different architectures for cpu and userspace
       "$ref": "#/definitions/architecture_data"
+  battery:
+    "$id": "/properties/battery"
+    oneOf:
+    - "$ref": "#/definitions/battery_data"
+    - descripion: Array of maps of model numbers to battery data
+      examples:
+      - SM-9000A: {removable: True, size: 1500, tech: 'Li-Ion'}
+      - SM-9000B:  {removable: False, size: 3500, tech: 'Li-Po'}
+      items:
+        additionalProperties:
+          "$ref": "#/definitions/battery_data"
+        description: Map of model number to battery_data
+        type: object
+      type: array
+    - enum:
+      - None
   bluetooth:
     "$id": "/properties/bluetooth"
     properties:
@@ -292,6 +329,7 @@ properties:
     type: array
 required:
 - architecture
+- battery
 - bluetooth
 - channels
 - codename
-- 
2.17.1

